tmt/steps/discover/__init__.py

```diff
@@ -36,7 +36,9 @@ class DiscoverStepData(tmt.steps.StepData):
         )
 
 
-class DiscoverPlugin(tmt.steps.GuestlessPlugin):
+# DiscoverFmf must be able to accept Execute step as its parent as a part
+# of the upgrade plugin, therefore Step is used here instead of Discover.
+class DiscoverPlugin(tmt.steps.GuestlessPlugin[tmt.steps.Step]):
```

Comment:
My two cents - I'd try to solve it in either `DiscoverFmf` or in `ExecuteUpgrade`, as it's clear just these two need to play nicely together, it's not worth the effort to spoil the whole `discover` just because of these two.

How about overriding the `step` in `DiscoverFmf`, allowing the special case?

```diff
diff --git a/tmt/steps/discover/__init__.py b/tmt/steps/discover/__init__.py
index 3b6143fc..06246db4 100644
--- a/tmt/steps/discover/__init__.py
+++ b/tmt/steps/discover/__init__.py
@@ -38,7 +38,7 @@ class DiscoverStepData(tmt.steps.StepData):
 
 # DiscoverFmf must be able to accept Execute step as its parent as a part
 # of the upgrade plugin, therefore Step is used here instead of Discover.
-class DiscoverPlugin(tmt.steps.GuestlessPlugin[tmt.steps.Step]):
+class DiscoverPlugin(tmt.steps.GuestlessPlugin['Discover']):
     """ Common parent of discover plugins """
 
     _data_class = DiscoverStepData
diff --git a/tmt/steps/discover/fmf.py b/tmt/steps/discover/fmf.py
index c43a3ee2..d7deb917 100644
--- a/tmt/steps/discover/fmf.py
+++ b/tmt/steps/discover/fmf.py
@@ -4,7 +4,7 @@ import os
 import re
 import shutil
 import subprocess
-from typing import Any, List, Optional, cast
+from typing import Any, List, Optional, Union, cast
 
 import click
 import fmf
@@ -16,6 +16,7 @@ import tmt.log
 import tmt.options
 import tmt.steps
 import tmt.steps.discover
+import tmt.steps.execute
 import tmt.utils
 from tmt.utils import Command, Path
 
@@ -171,6 +172,22 @@ class DiscoverFmf(tmt.steps.discover.DiscoverPlugin):
         "fmf-id",
         )
 
+    def __init__(
+            self,
+            *,
+            step: Union[tmt.steps.discover.Discover, tmt.steps.execute.Execute],
+            data: tmt.steps.StepData,
+            workdir: tmt.utils.WorkdirArgumentType = None,
+            logger: tmt.log.Logger) -> None:
+        super().__init__(
+            # ignore[arg-type]: we're allowing Execute on purpose, to accomodate the use of
+            # this plugin by `execute/upgrade`, where `upgrade` parent, i.e. `Execute`
+            # instance, serves as a parent.
+            step=step,  # type: ignore[arg-type]
+            data=data,
+            workdir=workdir,
+            logger=logger)
+
     @classmethod
     def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecoratorType]:
         """ Prepare command line options for given method """
```